sugar.go

```diff
@@ -212,25 +208,35 @@ func (s *SugaredLogger) sweetenFields(args []interface{}) []zapcore.Field {
 	if len(args) == 0 {
 		return nil
 	}
-	if len(args)%2 == 1 {
-		s.core.DPanic(_oddNumberErrMsg, Any("ignored", args[len(args)-1]))
-	}
 
-	fields := make([]zapcore.Field, len(args)/2)
-	for i := range fields {
-		keyIdx := 2 * i
-		val := args[keyIdx+1]
-		key, ok := args[keyIdx].(string)
+	// Allocate enough space for the common case; if users only pass structured
+	// fields, we'll need to grow the slice.
+	fields := make([]zapcore.Field, 0, len(args)/2)
+	var i int
+	for i < len(args) {
+		if f, ok := args[i].(zapcore.Field); ok {
+			fields = append(fields, f)
+			i++
+			continue
+		}
+		if i == len(args)-1 {
+			s.core.DPanic(_oddNumberErrMsg, Any("ignored", args[i]))
+			break
+		}
+		key, ok := args[i].(string)
+		val := args[i+1]
 		if !ok {
 			s.core.DPanic(
 				_nonStringKeyErrMsg,
-				Int("position", keyIdx),
-				Any("key", args[keyIdx]),
+				Int("position", i),
+				Any("key", args[i]),
 				Any("value", val),
 			)
-			key = fmt.Sprint(args[keyIdx])
+			key = fmt.Sprint(args[i])
```

Comment:
Is this the right form of degradation? A more conservative stance would be to `break`, so that we don't accidentally put a (maybe very large) value in as a "key". Basically, this case is similar to the parity fail case above, and I think we should "fail open" rather than "fail best effort" in production.